src/components/SingIn/SigIn.tsx

```diff
@@ -0,0 +1,115 @@
+import './SignIn.css';
+import { useState, useEffect } from 'react';
+import { authorize } from '../../utils/auth';
+import { setToken } from '../../utils/token';
+import { getUserInfo } from '../../utils/api';
+import { useNavigate } from 'react-router-dom';
+import { Link } from 'react-router-dom';
+import { setUser } from '../../redux/userSlice';
+import { useAppDispatch, useAppSelector } from '../../redux/hooks';
+import { getUser } from '../../redux/selectors';
+import Form from '../Form/Form';
+import { useFormAndValidation } from '../../hooks/useFormAndValidation';
+
+type FormData = {
+  email: string;
+  password: string;
+};
+
+function SingIn() {
+  const { values, handleChange, errors, isValid } =
+    useFormAndValidation<FormData>({
+      email: '',
+      password: '',
+    });
+  const navigate = useNavigate();
+  const dispatch = useAppDispatch();
+  const user = useAppSelector(getUser);
+  const [isLoading, setIsLoading] = useState(false);
+  const [formError, setFormError] = useState('');
+
+  const handleSubmit = (e: React.ChangeEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    setIsLoading(true);
+    handleLogin(values);
+  };
+
+  const handleLogin = (formData: FormData) => {
+    const { email, password } = formData;
+    if (!email || !password) {
+      return;
+    }
+    makeRequest(formData)
+      .then(() => navigate('/'))
+      .catch((error) => {
+        console.error(error);
+        setFormError(error.message || 'An unexpected error occurred.');
+      })
+      .finally(() => setIsLoading(false));
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(handleCloseModal) 
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js

  // here is an example
  const handleAddItem = (item) => {
    // here we create a function that returns a promise
    const makeRequest = () => {
      // `return` lets us use a promise chain `then, catch, finally`
      return addNewItem(item).then((item) => {
        setClothingItems([item, ...clothingItems]);
      });
    };
    // here we call handleSubmit passing the request
    handleSubmit(makeRequest);
  };

```

And for the 14th sprint:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```

So, this way you can remove a lot of code duplicating